mars/tensor/expressions/datasource/fromdataframe.py

```diff
@@ -0,0 +1,22 @@
+from .core import TensorHasInput
+from .... import opcodes as OperandDef
+from ....serialize import KeyField
+
+
+class TensorDataFrameDataSource(TensorHasInput):
+    """ represent tensor from DataFrame """
+
+    _op_type_ = OperandDef.TENSOR_FROM_DATAFRAME
+    _input = KeyField('_input')
+
+    def __init__(self, dtype=None, gpu=None, **kw):
+        super(TensorDataFrameDataSource, self).__init__(_dtype=dtype, _gpu=gpu,
+                                                        _sparse=True, **kw)
+
+
+def from_dataframe(in_df):
+    if not all(in_df.dtypes[0] == elem for elem in in_df.dtypes):
```

Comment:
In pandas, with heterogenous data, the lowest common type will have to be used. We need to follow this behavior.

Reply:
We can just create an empty pandas DataFrame, call `to_numpy` or `values` to get the promotion type.